sw/device/silicon_creator/rom/e2e/BUILD

```diff
@@ -1260,3 +1467,131 @@ test_suite(
         redact.lower(),
     ) for lc_state in structs.to_dict(CONST.LCV) for redact in REDACT],
 )
+
+# TODO: Define full list of GPRs that should be readable from debugger.
+ROM_E2E_DEBUG_GPRS = [
+    "a0",
+    "a1",
+    "a2",
+    "t0",
+    "t1",
+    "t2",
+]
+
+[
+    opentitan_gdb_fpga_cw310_test(
+        name = "rom_e2e_debug_test_otp_" + otp_name,
+        timeout = "short",
+        exit_success_pattern = "GDB-OK\r\nBFV:0142500d",
+        gdb_expect_output_sequence = [
+            ":::: GDB script complete.",
+            "Hart 0 unexpectedly reset!",
+        ],
+        gdb_script = """
+            define assert_pc
+                printf ":::: PC=%p. Expected PC=%p.\\n", $pc, $arg0
+                if $pc != $arg0
+                    quit 1
+                end
+            end
+
+            target extended-remote :3333
+
+            echo :::: Send OpenOCD the 'reset halt' command.\\n
+            monitor reset halt
+
+            echo :::: Load ROM symbols into GDB.\\n
+            file rom.elf
+
+            # Immediately after reset, we should be sitting on the Ibex reset
+            # handler, the 33rd entry in the interrupt vector.
+            set $ibex_reset_handler = _rom_interrupt_vector_asm + 32 * 4
+            assert_pc $ibex_reset_handler
+
+            # Executing the `j _rom_start_boot` instruction should take us
+            # to _rom_start_boot.
+            stepi
+            assert_pc _rom_start_boot
+
+            echo :::: Run until we check whether ROM execution is enabled.\\n
+            break kRomStartBootMaybeHalt
+            continue
+            assert_pc kRomStartBootMaybeHalt
+
+            # Use three of four hardware breakpoints. Leave one unused so we can
+            # use `call` and `finish`. (See RvCoreIbexDbgHwBreakNum in
+            # hw/top_earlgrey/data/top_earlgrey.hjson.)
+            echo :::: Set all but one available breakpoints.\\n
+            break rom_main
+            break rom_init
+
+            echo :::: Pretend execution is enabled.\\n
+            set $pc = kRomStartBootExecEn
+
+            echo :::: Continue until rom_main.\\n
+            continue
+            assert_pc rom_main+0x18
```

Comment:
@alphan Despite setting the breakpoint on `rom_main`, we consistently land 0x18 bytes past the symbol. Any idea what's going on?

For reference:

```
00008800 <rom_main>:
rom_main():
/proc/self/cwd/sw/device/silicon_creator/rom/rom.c:438
void rom_main(void) {
    8800:                 1101                  addi    sp,sp,-32
    8802:                 ce06                  sw      ra,28(sp)
    8804:                 cc22                  sw      s0,24(sp)
    8806:                 ca26                  sw      s1,20(sp)
    8808:                 c84a                  sw      s2,16(sp)
    880a:                 c64e                  sw      s3,12(sp)
    880c:                 c452                  sw      s4,8(sp)
    880e:                 c256                  sw      s5,4(sp)
    8810:                 c0418a93              addi    s5,gp,-1020 # 100023cc <rom_counters>
    8814:                 1a500513              li      a0,421
/proc/self/cwd/sw/device/silicon_creator/rom/rom.c:439
  CFI_FUNC_COUNTER_INIT(rom_counters, kCfiRomMain);
    8818:                 00aaa023              sw      a0,0(s5)
/proc/self/cwd/sw/device/silicon_creator/rom/rom.c:441
  CFI_FUNC_COUNTER_PREPCALL(rom_counters, kCfiRomMain, 1, kCfiRomInit);
    881c:                 000aa583              lw      a1,0(s5)
    8820:             /-- 00a58663              beq     a1,a0,882c <rom_main+0x2c>
    8824:             |   0000                  unimp
    8826:             |   0000                  unimp
    8828:             |   0000                  unimp
    882a:             |   0000                  unimp
    882c:             \-> 000aa503              lw      a0,0(s5)
    8830:                 05a50513              addi    a0,a0,90
    8834:                 00aaa023              sw      a0,0(s5)
```

Reply:
I'm not sure, is this a fixed constant for every symbol or just `rom_main`? Is there any chance gdb is looking at an out-of-date symbol file? @GregAC any ideas?

Reply:
Well, it's not a fixed constant for every symbol, but `rom_main` is the only C function I've set a breakpoint for *and* asserted the PC value.

The symbol file should be current because I'm using `--define bitstream=gcp_splice`.

Reply:
This appears to be a feature of GDB that we hit. It tries to automatically skip the function prologue. I have a workaround :) https://sourceware.org/gdb/wiki/Internals/Prologue%20Analysis